### PR TITLE
vm: split up `TCtx`

### PR DESCRIPTION
**Pull request by zerbina**
This is a sketch of how I plan to split up the state stored as part of `TCtx` into multiple smaller pieces.

As for the implementation, the plan is to, at first, use accessor templates mapping the old fields to their new counterparts and then incrementally replace usage of `TCtx`.

**Motivations**:
- make it more clear what code uses/modifies which state
- make the VM code more modular
- more self-documenting implementation
- prepare for the adjustments required for the planned `vmgen`-independent execution

---

## Notes for Reviewers
* the names of fields and types are not final. If you have any suggestions, please state them

### CODE REVIEW

File: compiler/vm/vmdef.nim [hunk 1]
```diff
@@ -426,9 +427,12 @@ type
...
 
-  TypeInfoCache* = object
-    ## An append-only cache for everything type related
-    lut*: Table[ItemId, PVmType] ## `PType`-id -> `PVmType` mappings
+  TypeInfoCache* = TypeGenState ## transition helper
+  TypeGenState* = object
+    ## All the state necessary for incrementally generating `VmType`s from
+    ## `PType`s
+    lut*: Table[ItemId, PVmType] ##
+      ## `PType`-id -> `PVmType` mappings. Also acts as a cache
```
Comment (saem) [inline compiler/vm/vmdef.nim:435]:
```suggestion
      ## Look-Up Table: `PType`-id -> `PVmType` mappings. Also acts as a cache
```

File: compiler/vm/vmdef.nim [hunk 2]
```diff
@@ -460,77 +470,110 @@ type
...
-  #      - 'execution environment': types, globals, constants, functions, etc.;
-  #        populated by code-gen (vmgen) or the compilerapi and possibly
-  #        reused across compiler invocations (relevant for IC). Not mutated
-  #        by the execution engine
-  #      - code and debug information
-  #      - 'vmgen' state: auxiliary data used during code generation, reused
-  #        across vmgen invocations for efficiency
+  ExecState* = object
+    ## The state that makes up a single VM invocation. Mutated by execution
+    sframes*: seq[TStackFrame] ## The stack of the currently running code # XXX: rename to `stack`?
```
Comment (saem) [inline compiler/vm/vmdef.nim:481]: I'd say `callStack` personally.

File: compiler/vm/vmdef.nim [hunk 3]
```diff
@@ -460,77 +470,110 @@ type
...
 
-  PCtx* = ref TCtx
-  TCtx* = object of TPassContext
-    # XXX: TCtx stores three different things:
-    #  - VM execution state
-    #  - VM environment (code, constants, type data, etc.)
-    #  - vmgen context/state
-    # These should be encapsulated into standalone types
+    currentExceptionA*, currentExceptionB*: HeapSlotHandle
```
Comment (saem) [inline compiler/vm/vmdef.nim:483]: Always thought it was weird that we'd track two exceptions, but I didn't get a chance to dig into why it was necessary.

File: compiler/vm/vmdef.nim [hunk 4]
```diff
@@ -460,77 +470,110 @@ type
...
 
-    code*: seq[TInstr]
-    debug*: seq[TLineInfo]  # line info for every instruction; kept separate
-                            # to not slow down interpretation
-    sframes*: seq[TStackFrame] ## The stack of the currently running code # XXX: rename to `stack`?
-    globals*: seq[HeapSlotHandle] ## Stores each global's corresponding heap slot
-    constants*: seq[VmConstant] ## constant data
+    loopIterations*: int
+
+  ExecStateExtra* = object
```
Comment (saem) [inline compiler/vm/vmdef.nim:488]: CompTimeState

File: compiler/vm/vmdef.nim [hunk 5]
```diff
@@ -460,77 +470,110 @@ type
...
-    rtti*: seq[VmTypeInfo] ## run-time-type-information as needed by
-                           ## conversion and `repr`
-    functions*: seq[VmFunctionObject] ## all functions known to the VM. Indexed
-                                      ## by `FunctionIndex`
+    module*: PSym
+    cache*: IdentCache
+    graph*: ModuleGraph
+    templInstCounter*: ref int # gives every template instantiation a unique ID, needed here for getAst
+
+  SharedExecState* = object
```
Comment (saem) [inline compiler/vm/vmdef.nim:501]: `HostExecState`?

I personally was thinking of using a hierarchy for compiler config such as: Host -> Env -> Compiler -> Project; these aren't finalized or anything.

When thinking about the VM right now the first thing that came to mind: Sys -> VMHost -> VM -> Guest

File: compiler/vm/vmdef.nim [hunk 6]
```diff
@@ -460,77 +470,110 @@ type
...
-    # XXX: could and should be merged with `procToCodePos` table
-    procToFuncObj*: Table[int, FunctionIndex] ## symbol id -> index into `functions`
+    profiler*: Profiler
 
-    currentExceptionA*, currentExceptionB*: HeapSlotHandle
-    exceptionInstr*: int # index of instruction that raised the exception
-    prc*: PProc
-    module*: PSym
-    callsite*: PNode
+  ExecEnv* = object
```
Comment (saem) [inline compiler/vm/vmdef.nim:513]: `CompTimeEnv`, assuming consistency with the earlier CompTime suggestion.

File: compiler/vm/vmdef.nim [hunk 7]
```diff
@@ -460,77 +470,110 @@ type
...
+    types*: Types
+    rtti*: seq[VmTypeInfo]
+
+    constants*: seq[VmConstant] ## constant data
+    functions*: seq[VmFunctionObject] ## all functions known to the VM. Indexed
+                                      ## by `FunctionIndex`
+
+    callbacks*: seq[tuple[key: string, value: VmCallback]]
+
+    # XXX: maybe split these off into a separate object (e.g. ExecConfig)?
```
Comment (saem) [inline compiler/vm/vmdef.nim:527]: I've been reading down the code, so evolving my thinking as I'm going. `Exec` might be good for single execution of the VM?

splitting out ExecConfig makes sense, I think.

**Review by saem — COMMENTED** [review]
Added a few thoughts.

File: compiler/vm/vmdef.nim (hunk 3 above)
Comment (zerbina) [reply compiler/vm/vmdef.nim:483]: `currentExceptionB` is never used and from my last look at git history, IIRC, it also never was. I just haven't gotten around to removing it yet.

File: compiler/vm/vmdef.nim (hunk 6 above)
Comment (zerbina) [reply compiler/vm/vmdef.nim:513]: `ExecEnv` would also be available during standalone execution whereas `CompTimeState` would not, so I don't think `CompTime` fits here

**Comment by zerbina** [comment]
This PR is stale and will not be worked on anymore. I'm keeping it open in order to highlight that splitting up `TCtx` is still an ongoing project.

**Comment by zerbina** [comment]
Superseded by many smaller split-outs, with #843 and #774 being the most significant ones.

**Pull request closed without merging**